File: projJava/src/Model/InfoFatur.java
package Model;

import java.io.Serializable;

public class InfoFatur implements IInfoFatur, Serializable {
    private static final long serialVersionUID = -78465562523422L;
    private final String produto;
    private final int mes;
    private final int filial;
    private int nr_vendasN;  //número de clientes diferentes que compraram este produto em N
    private int nr_vendasP;  //número de clientes diferentes que compraram este produto em P
    private double fatP;     //total da faturacao em modo P
    private double fatN;     //total da faturacao em modo N
    private int unidT;       //número Total de unidades adquiridas deste produto

    /**
     * Construtor vazio
     */
    public InfoFatur(){
        this.produto = "";
        this.mes = 0;
        this.filial = 0;
        this.nr_vendasN = 0;
        this.nr_vendasP = 0;
        this.fatP = 0.0;
        this.fatN = 0.0;
        this.unidT = 0;
    }

    /**
     * Construtor de uma infofatur, a partir de uma linha do ficheiro
     * @param venda String com informação da venda
     */
    public InfoFatur (String venda){
        String[] vendaSplit = venda.split(" ");
        this.produto = vendaSplit[0];
        this.mes = Integer.parseInt(vendaSplit[5]);
        this.filial = Integer.parseInt(vendaSplit[6]);
        if((vendaSplit[3].charAt(0))=='N'){
            this.nr_vendasN++;
            this.fatN += Double.parseDouble(vendaSplit[1]);
        }
        else {
            this.nr_vendasP++;
            this.fatP += Double.parseDouble(vendaSplit[1]);
        }
        this.unidT += Integer.parseInt(vendaSplit[2]);
    }

    /**
     * Vê se uma infofatur tem igual produto, mes e filial
     * @param aux infofatur
     * @return boolean
     */
    public boolean compareInfos(IInfoFatur aux){
        if(this.get_produto().equals(aux.get_produto())){
            if(this.get_mes() == (aux.get_mes())){
                return this.get_fil() == (aux.get_fil());
            }
        }
        return false;
    }

    /**
     * Soma duas infofatur
     * @param soma infofatur a somar
     */
    public void addInfos(IInfoFatur soma){
        this.nr_vendasN += soma.get_nr_vendasN();
        this.nr_vendasP += soma.get_nr_vendasP();
        this.fatN += soma.get_FatN();
        this.fatP += soma.get_FatP();
        this.unidT += soma.get_unidT();
    }

    /**
     * Getter do codigo do produto
     * @return Codigo do produto
     */
    public String get_produto(){
        return this.produto;
    }

    /**
     * Getter do mês da infofatur
     * @return mês da infofatur
     */
    public int get_mes(){
        return this.mes;
    }

    /**
     * Getter da filial da infofatur
     * @return filial da infofatur
     */
    public int get_fil(){
        return this.filial;
    }

    /**
     * Getter do número de vendas do tipo N da infofatur
     * @return número de vendas do tipo N
     */
    public  int get_nr_vendasN(){
        return this.nr_vendasN;
    }

    /**
     * Getter do número de vendas do tipo P da infofatur
     * @return número de vendas do tipo P
     */
    public int get_nr_vendasP(){
        return this.nr_vendasP;
    }

    /**
     * Getter do total faturado do tipo N da infofatur
     * @return total faturado do tipo N
     */
    public double get_FatN(){
        return this.fatN;
    }

    /**
     * Getter do total faturado do tipo P da infofatur
     * @return total faturado do tipo P
     */
    public double get_FatP(){
        return this.fatP;
    }

    /**
     * Getter do total de unidades da infofatur
     * @return total de unidades
     */
    public int get_unidT(){
        return this.unidT;
    }


}

File: projC/src/InfoFatur.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../include/InfoFatur.h"

/** Módulo responsável pela criação da estrutura que irá ser utilizada para preencher a AVL da Faturação Global.*/

/** Esta struct armazena certas informações adquiridas através da leitura de uma linha de venda válida.Portanto, guarda os seguintes campos:
* 1.Código do produto (de modo a sabermos a que node correspondente podemos adicionar esta Info_fatur);
* 2.Mês de compra;
* 3.Filial;
* 4.Número de vendas deste produto em modo N;
* 5.Número de vendas deste produto em modo P;
* 6.Número de vendas no total, isto é N+P;
* 7.Faturação Global deste produto em modo P;
* 8.Faturação Global deste produto em modo N;
* 9.Faturação Global no total;
* 10.Unidades adquiridas deste produto no total.
*/
struct info_fatur{
	char produto[40];
	int mes;
	int fil;
	int nr_vendasN;
	int nr_vendasP;
	int nr_vendasTotal;
	double fatP;
	double fatN;
	double fatTotal;
	int unidT;
};

/** Inicialização da struct Info_fatur, com maior parte dos campos a 0.
*	NOTA: produtos que não foram comprados vão ter estes campos a 0. */
Info_fatur init_infoFatur(){ 
	Info_fatur p = (Info_fatur) calloc(1,sizeof(struct info_fatur));
	p->mes = 0;
	p->fil = 0;
	p->nr_vendasN = 0;
	p->nr_vendasP = 0;
	p->fatP = 0;
	p->fatN = 0;
	p->nr_vendasTotal = 0;
	p->fatTotal = 0;
	p->unidT = 0;

	return p;
}

/** Inicialização da struct Info_fatur mas com o campo do código do produto já preenchido. Usamos esta função para adicionar apenas os produtos do catálogo à AVL
* da faturação mas os campos da Info_Fatur vazios pois nesta fase ainda não temos acesso às vendas realizadas de cada produto. */
Info_fatur init_infoFatur_Prod(const char *prod){
	Info_fatur p = init_infoFatur();
	strcpy(p->produto,prod);

	return p;
}

/** Liberta a memória da struct*/
void destroy_infoFatur (Info_fatur p) {
	free(p);
}

/** Quando já temos acesso às informações de cada venda, e sabemos que a venda é válida, adicionamos os dados da linha de venda à info_fatur do produto em causa*/
Info_fatur insert_infoFatur(Info_fatur p, char* pd, char unid[],char pr[] ,char tipo[] ,char mes[],char fil[]){ 
	strcpy(p->produto,pd);
	if(strcmp("N",tipo) == 0){
		(p->fatN) += (atof(pr)*atoi(unid));
		(p->nr_vendasN)++;
	}
	else {
		(p->fatP) += (atof(pr)*atoi(unid));
		(p->nr_vendasP)++;
	}
	(p->mes) = atoi(mes);
	(p->fil) = atoi(fil);
	(p->fatTotal) = (p->fatP) + (p->fatN);
	(p->nr_vendasTotal)++;
	(p->unidT) = atoi(unid);
	return p;
}

/** Compara duas info_faturs's, caso ambas tenham informações sobre o mesmo produto, iremos verificar se essas informações são do mesmo mês e da mesma filial,
* se isto se confirmar então o return será 0, caso contrário devolve 1. Esta função será utilizada para depois adicionar as informações de duas info_faturs para
* os casos em que esta função devolve 0.*/
int compare(Info_fatur p, Info_fatur f){ 
	if(strcmp(p->produto,f->produto)==0){
		if(p->mes==f->mes){
			if(p->fil==f->fil){
				return 0;
			}
		}
	}
	return 1;
}

/** Adiciona as informações de duas Info_Fatur's. */
Info_fatur add_information(Info_fatur fat,Info_fatur info){ 
	fat->nr_vendasTotal += info->nr_vendasTotal;
	fat->nr_vendasP += info->nr_vendasN;
	fat->nr_vendasN += info->nr_vendasP;
	fat->fatN += info->fatN;
	fat->fatP += info->fatP;
	fat->fatTotal += info->fatTotal;
	fat->unidT += info->unidT;
	return fat;
}

/** Funções do tipo "get" que nos dão acesso aos campos desta struct noutros módulos. */
char* get_prodID(Info_fatur fat){
	return(fat->produto);
}

int get_mesFt(Info_fatur fat){
	return(fat->mes);
}

int get_filFt(Info_fatur fat){
	return(fat->fil);
}

int get_nr_vendasT(Info_fatur fat){
	return(fat->nr_vendasTotal);
}

int get_nr_vendasN(Info_fatur fat){
	return(fat->nr_vendasN);
}

int get_nr_vendasP(Info_fatur fat){
	return(fat->nr_vendasP);
}

double get_FatN(Info_fatur fat){
	return(fat->fatN);
}

double get_FatP(Info_fatur fat){
	return(fat->fatP);
}

double get_FatTotal(Info_fatur fat){
	return(fat->fatTotal);
}

int get_unidT(Info_fatur fat){
	return(fat->unidT);
}
